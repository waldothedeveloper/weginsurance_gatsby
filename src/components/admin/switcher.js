import PropTypes from "prop-types";
import React from "react";
import { classNames } from "../../utils/classNames";

export const Switcher = ({ handleModule, module }) => (
  <>
    {module.map((item) => (
      <button
        key={item.name}
        onClick={() => handleModule(item.name)}
        className={classNames(
          item.current
            ? `bg-sky-800 text-white`
            : `text-white hover:bg-sky-600`,
          `group flex w-full items-center rounded-md px-2 py-2 text-base font-medium`
        )}
      >
        <item.icon
          className="mr-4 h-6 w-6 flex-shrink-0 text-sky-300"
          aria-hidden="true"
        />
        {item.name}
      </button>
    ))}
  </>
);

Switcher.propTypes = {
  handleModule: PropTypes.func.isRequired,
  module: PropTypes.array.isRequired,
};
