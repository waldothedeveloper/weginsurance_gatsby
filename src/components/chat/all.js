import React from "react";

export const Chat = () => (
  <div>
    <h1>This is the CHAT component</h1>
  </div>
);
