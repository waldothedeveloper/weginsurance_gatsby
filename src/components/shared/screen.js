import React from "react";
import propTypes from "prop-types";
export const SharedScreen = ({
  rightPanel: RightPanel,
  leftPanel: LeftPanel,
}) => (
  <div className="z-0 mx-auto grid grid-cols-12 gap-8 px-6">
    <div className="col-span-4 h-full rounded-lg border-2 border-solid border-gray-100 py-6 lg:px-4">
      {LeftPanel}
    </div>
    <div className="col-span-8 h-full rounded-lg border-2 border-solid border-gray-100 pt-6 lg:px-1">
      {RightPanel}
    </div>
  </div>
);

SharedScreen.propTypes = {
  rightPanel: propTypes.node.isRequired,
  leftPanel: propTypes.node.isRequired,
};
