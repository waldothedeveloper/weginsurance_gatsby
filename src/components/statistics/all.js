import React from "react";
export const Statistics = () => <div>I am the Statistics Component</div>;
