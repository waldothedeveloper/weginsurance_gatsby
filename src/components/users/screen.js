import { EditUserForm } from "./editUserForm";
import { EmptyUserList } from "./emptyUserList";
import { NewUserForm } from "./newUserForm";
import { NoUsers } from "./empty";
import React from "react";
import { UserDetails } from "./details";
import { UserScreenLoader } from "../shared/loaders/usersScreenLoader";
import { UsersDictionary } from "./usersDictionary";
import { useCreateUser } from "../../hooks/user/useCreateUser";
import { useEditUser } from "../../hooks/user/useEditUser";
import { useReadUsers } from "../../hooks/user/useReadUsers";
import { useSwitcherUserCRUDMode } from "../../hooks/user/useSwitcherUserCRUDMode";

export const UsersScreen = () => {
  const { users } = useReadUsers();

  const {
    selectedUser,
    editUser,
    newUser,
    handleEditUser,
    handleNewUser,
    handleSelectUser,
    handleUserCRUDAction,
    handleUpdatedUser,
  } = useSwitcherUserCRUDMode();

  const { values, handleChange, handleSubmit, errors } = useCreateUser();
  const {
    userToEdit,
    editUserErrors,
    handleEditUserChange,
    handleUserEditSubmit,
  } = useEditUser(selectedUser, handleUpdatedUser);

  return (
    <div className="z-0 mx-auto grid grid-cols-12 gap-8 px-6">
      <div className="col-span-4 h-screen overflow-y-auto rounded-lg border-2 border-solid border-gray-100 py-6 lg:px-4">
        {users && users.length > 0 ? (
          <UsersDictionary
            users={users}
            handleUser={handleNewUser}
            handleSelectUser={handleSelectUser}
          />
        ) : users && users.length === 0 ? (
          <EmptyUserList createUser={handleNewUser} />
        ) : (
          <UserScreenLoader />
        )}
      </div>
      <div className="col-span-8 h-screen overflow-y-auto rounded-lg border-2 border-solid border-gray-100 pb-40 lg:px-4">
        {newUser ? (
          <NewUserForm
            values={values}
            handleChange={handleChange}
            handleSubmit={handleSubmit}
            errors={errors}
            handleUser={handleUserCRUDAction}
          />
        ) : selectedUser ? (
          <UserDetails
            selectedUser={selectedUser}
            handleSelectedUser={handleSelectUser}
            handleEditUser={handleEditUser}
            users={users}
            handleUser={handleUserCRUDAction}
          />
        ) : editUser ? (
          <EditUserForm
            values={userToEdit}
            handleChange={handleEditUserChange}
            handleSubmit={handleUserEditSubmit}
            errors={editUserErrors}
            handleUser={handleUserCRUDAction}
          />
        ) : (
          <NoUsers createUser={handleNewUser} />
        )}
      </div>
    </div>
  );
};
