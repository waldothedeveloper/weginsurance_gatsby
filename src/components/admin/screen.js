import React, { useEffect, useState } from "react";

import { Chat } from "../chat/index";
import PropTypes from "prop-types";
import { Statistics } from "../statistics/all";
import { UsersScreen } from "../users/index";

const Skeleton = ({ children }) => <main className="py-10">{children}</main>;

export const Screen = ({ module }) => {
  const [componentToRender, setComponentToRender] = useState(`Dashboard`);

  useEffect(() => {
    const currentModule = module.filter((item) => item.current)[0].name;
    setComponentToRender(currentModule);
  }, [module]);

  switch (componentToRender) {
    case `Dashboard`:
      return (
        <Skeleton>
          <Statistics />
        </Skeleton>
      );
    case `Clientes`:
      return (
        <Skeleton>
          <UsersScreen />
        </Skeleton>
      );
    case `Chat`:
      return (
        <Skeleton>
          <Chat />
        </Skeleton>
      );
    default:
      return (
        <Skeleton>
          <Statistics />
        </Skeleton>
      );
  }
};

Screen.propTypes = {
  module: PropTypes.array.isRequired,
};

Skeleton.propTypes = {
  children: PropTypes.node.isRequired,
};
