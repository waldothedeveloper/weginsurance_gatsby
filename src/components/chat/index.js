import { ChatWall } from "./wall";
import { Error } from "./error";
import { NoUserSelected } from "./noUserSelected";
import React from "react";
import { SharedScreen } from "../shared/screen";
import { Spinner } from "../shared/spinner";
import { Transition } from "@headlessui/react";
import { UserScreenLoader } from "../shared/loaders/usersScreenLoader";
import { UsersList } from "./usersList";
import { useConfigureChat } from "../../hooks/chat/useConfigureChat";
import { useReadUsers } from "../../hooks/user/useReadUsers";
import { useSelectUserToChat } from "../../hooks/chat/useSelectUserToChat";

//
export const Chat = () => {
  const { users } = useReadUsers();
  const { participantInfo, handleParticipantInfo } = useSelectUserToChat();

  const { statusString, messages, newMessage, handleChange, handleSubmit } =
    useConfigureChat(participantInfo);

  return (
    <SharedScreen
      leftPanel={
        <Transition
          show={true}
          enter="transition-opacity duration-75"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="transition-opacity duration-150"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          {users && users.length > 0 ? (
            <div className="flex flex-1 flex-col overflow-hidden">
              <UsersList
                users={users}
                chooseParticipant={handleParticipantInfo}
              />
            </div>
          ) : users && users.length === 0 ? (
            <div>WE NEED SOMETHING HERE</div>
          ) : (
            <UserScreenLoader chat={true} />
          )}
        </Transition>
      }
      rightPanel={
        statusString === `You are connected.` ? (
          <ChatWall
            messages={messages}
            handleChange={handleChange}
            handleSubmit={handleSubmit}
            newMessage={newMessage}
          />
        ) : statusString === `Connecting to Twilio…` ? (
          <Spinner>Conectando el chat...</Spinner>
        ) : statusString === `Failed to connect.` ? (
          <Error />
        ) : statusString === `Disconnecting from Twilio…` ? (
          <Spinner>Desconectando el chat...</Spinner>
        ) : (
          <NoUserSelected />
        )
      }
    />
  );
};
