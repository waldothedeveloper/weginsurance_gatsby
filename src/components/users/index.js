import { EditUserForm } from "./editUserForm";
import { EmptyUserList } from "./emptyUserList";
import { NewUserForm } from "./newUserForm";
import { NoUsers } from "./empty";
import React from "react";
import { SharedScreen } from "../shared/screen";
import { Transition } from "@headlessui/react";
import { UserDetails } from "./details";
import { UserScreenLoader } from "../shared/loaders/usersScreenLoader";
import { UsersDictionary } from "./usersDictionary";
import { useCreateUser } from "../../hooks/user/useCreateUser";
import { useEditUser } from "../../hooks/user/useEditUser";
import { useReadUsers } from "../../hooks/user/useReadUsers";
import { useSwitcherUserCRUDMode } from "../../hooks/user/useSwitcherUserCRUDMode";

export const UsersScreen = () => {
  const { users } = useReadUsers();

  const {
    selectedUser,
    editUser,
    newUser,
    handleEditUser,
    handleNewUser,
    handleSelectUser,
    handleUserCRUDAction,
    handleUpdatedUser,
  } = useSwitcherUserCRUDMode();

  const { values, handleChange, handleSubmit, errors } = useCreateUser();
  const {
    userToEdit,
    editUserErrors,
    handleEditUserChange,
    handleUserEditSubmit,
  } = useEditUser(selectedUser, handleUpdatedUser);

  return (
    <SharedScreen
      leftPanel={
        <Transition
          show={true}
          enter="transition-opacity duration-75"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="transition-opacity duration-150"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          {users && users.length > 0 ? (
            <UsersDictionary
              users={users}
              handleUser={handleNewUser}
              handleSelectUser={handleSelectUser}
            />
          ) : users && users.length === 0 ? (
            <EmptyUserList createUser={handleNewUser} />
          ) : (
            <UserScreenLoader />
          )}
        </Transition>
      }
      rightPanel={
        <Transition
          show={true}
          enter="transition-opacity duration-75"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="transition-opacity duration-150"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          {newUser ? (
            <div className="h-[82vh] flex-none overflow-y-scroll px-2">
              <NewUserForm
                values={values}
                handleChange={handleChange}
                handleSubmit={handleSubmit}
                errors={errors}
                handleUser={handleUserCRUDAction}
              />
            </div>
          ) : selectedUser ? (
            <UserDetails
              selectedUser={selectedUser}
              handleSelectedUser={handleSelectUser}
              handleEditUser={handleEditUser}
              users={users}
              handleUser={handleUserCRUDAction}
            />
          ) : editUser ? (
            <div className="h-[82vh] flex-none overflow-y-scroll px-2">
              <EditUserForm
                values={userToEdit}
                handleChange={handleEditUserChange}
                handleSubmit={handleUserEditSubmit}
                errors={editUserErrors}
                handleUser={handleUserCRUDAction}
              />
            </div>
          ) : (
            <NoUsers createUser={handleNewUser} />
          )}
        </Transition>
      }
    />
  );
};
