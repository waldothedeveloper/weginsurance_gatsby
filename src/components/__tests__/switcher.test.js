import { FolderIcon, HomeIcon, UsersIcon } from "@heroicons/react/outline";
import { render, screen } from "@testing-library/react";

import { Chat } from "../../components/chat/all";
import React from "react";
import { Statistics } from "../../components/statistics/all";
import { Switcher } from "../admin/switcher";
import { UsersScreen } from "../../components/users/screen";
import userEvent from "@testing-library/user-event";

//
const user = userEvent.setup();
const handleModule = jest.fn();
const module = [
  {
    name: `Dashboard`,
    icon: HomeIcon,
    current: true,
    comp: Statistics,
  },
  {
    name: `Clientes`,
    icon: UsersIcon,
    current: false,
    comp: UsersScreen,
  },
  { name: `Chat`, icon: FolderIcon, current: false, comp: Chat },
];

//
describe(`Switcher`, () => {
  it(`should render correctly`, () => {
    const { container } = render(
      <Switcher handleModule={handleModule} module={module} />
    );
    expect(container).toMatchSnapshot();
  });

  it(`should call handleModule on click`, async () => {
    render(<Switcher handleModule={handleModule} module={module} />);
    await user.click(screen.getByText(`Clientes`));
    expect(handleModule).toHaveBeenCalled();
  });
});
