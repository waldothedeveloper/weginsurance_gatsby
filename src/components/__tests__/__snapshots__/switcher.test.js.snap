// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher should render correctly 1`] = `
<div>
  <button
    class="bg-sky-800 text-white group flex w-full items-center rounded-md px-2 py-2 text-base font-medium"
  >
    <svg
      aria-hidden="true"
      class="mr-4 h-6 w-6 flex-shrink-0 text-sky-300"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    Dashboard
  </button>
  <button
    class="text-white hover:bg-sky-600 group flex w-full items-center rounded-md px-2 py-2 text-base font-medium"
  >
    <svg
      aria-hidden="true"
      class="mr-4 h-6 w-6 flex-shrink-0 text-sky-300"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 4.354a4 4 0 110 5.292M15 21H3v-1a6 6 0 0112 0v1zm0 0h6v-1a6 6 0 00-9-5.197M13 7a4 4 0 11-8 0 4 4 0 018 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    Clientes
  </button>
  <button
    class="text-white hover:bg-sky-600 group flex w-full items-center rounded-md px-2 py-2 text-base font-medium"
  >
    <svg
      aria-hidden="true"
      class="mr-4 h-6 w-6 flex-shrink-0 text-sky-300"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 7v10a2 2 0 002 2h14a2 2 0 002-2V9a2 2 0 00-2-2h-6l-2-2H5a2 2 0 00-2 2z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    Chat
  </button>
</div>
`;
